File: shared/errors/app_error.dart
import 'dart:math';

/// Generate a random trace ID for error tracking
String generateTraceId() {
  final random = Random();
  final characters = 'abcdefghijklmnopqrstuvwxyz0123456789';
  
  return List.generate(
    16,
    (index) => characters[random.nextInt(characters.length)],
  ).join();
}

/// Application error class with standardized structure
class AppError implements Exception {
  final String code;
  final String message;
  final Map<String, dynamic>? details;
  final String traceId;
  final DateTime timestamp;

  AppError({
    required this.code,
    required this.message,
    this.details,
    String? traceId,
  })  : traceId = traceId ?? generateTraceId(),
        timestamp = DateTime.now();

  /// Convert error to JSON format
  Map<String, dynamic> toJson() {
    return {
      'error': {
        'code': code,
        'message': message,
        'details': details,
        'timestamp': timestamp.toIso8601String(),
        'trace_id': traceId,
      }
    };
  }

  /// Create AppError from JSON
  factory AppError.fromJson(Map<String, dynamic> json) {
    final errorData = json['error'] as Map<String, dynamic>;
    
    return AppError(
      code: errorData['code'] as String,
      message: errorData['message'] as String,
      details: errorData['details'] as Map<String, dynamic>?,
      traceId: errorData['trace_id'] as String?,
    );
  }

  /// Convert to string representation
  @override
  String toString() {
    return 'AppError(code: $code, message: $message, traceId: $traceId)';
  }

  /// Create a copy with optional modifications
  AppError copyWith({
    String? code,
    String? message,
    Map<String, dynamic>? details,
    String? traceId,
  }) {
    return AppError(
      code: code ?? this.code,
      message: message ?? this.message,
      details: details ?? this.details,
      traceId: traceId ?? this.traceId,
    );
  }
}

/// Common error codes for the application
class ErrorCodes {
  // Trading errors (001-099)
  static const String tradingInsufficientBalance = 'TRADING_001';
  static const String tradingInvalidSymbol = 'TRADING_002';
  static const String tradingInvalidQuantity = 'TRADING_003';
  static const String tradingInvalidPrice = 'TRADING_004';
  static const String tradingOrderNotFound = 'TRADING_005';
  static const String tradingMarketClosed = 'TRADING_006';
  static const String tradingPositionNotFound = 'TRADING_007';
  static const String tradingRiskLimitExceeded = 'TRADING_008';
  
  // Authentication errors (100-199)
  static const String authInvalidCredentials = 'AUTH_100';
  static const String authTokenExpired = 'AUTH_101';
  static const String authTokenInvalid = 'AUTH_102';
  static const String authPermissionDenied = 'AUTH_103';
  static const String authSignatureInvalid = 'AUTH_104';
  static const String authNonceInvalid = 'AUTH_105';
  
  // Blockchain errors (200-299)
  static const String blockchainConnectionFailed = 'BLOCKCHAIN_200';
  static const String blockchainTransactionFailed = 'BLOCKCHAIN_201';
  static const String blockchainInsufficientGas = 'BLOCKCHAIN_202';
  static const String blockchainContractError = 'BLOCKCHAIN_203';
  static const String blockchainInvalidAddress = 'BLOCKCHAIN_204';
  static const String blockchainNetworkError = 'BLOCKCHAIN_205';
  
  // RAG/Knowledge base errors (300-399)
  static const String ragIndexNotFound = 'RAG_300';
  static const String ragSearchFailed = 'RAG_301';
  static const String ragDocumentNotFound = 'RAG_302';
  static const String ragProcessingFailed = 'RAG_303';
  static const String ragDatabaseError = 'RAG_304';
  
  // API/Network errors (400-499)
  static const String apiNetworkError = 'API_400';
  static const String apiTimeoutError = 'API_401';
  static const String apiRateLimitExceeded = 'API_402';
  static const String apiInvalidRequest = 'API_403';
  static const String apiInvalidResponse = 'API_404';
  static const String apiServiceUnavailable = 'API_405';
  
  // System errors (500-599)
  static const String systemConfigurationError = 'SYSTEM_500';
  static const String systemDatabaseError = 'SYSTEM_501';
  static const String systemFileError = 'SYSTEM_502';
  static const String systemMemoryError = 'SYSTEM_503';
  static const String systemUnknownError = 'SYSTEM_504';
}

/// Trading-specific error classes
class TradingError extends AppError {
  TradingError({
    required String code,
    required String message,
    Map<String, dynamic>? details,
    String? traceId,
  }) : super(
          code: code,
          message: message,
          details: details,
          traceId: traceId,
        );

  factory TradingError.insufficientBalance({
    required double required,
    required double available,
  }) {
    return TradingError(
      code: ErrorCodes.tradingInsufficientBalance,
      message: 'Insufficient balance for trade',
      details: {
        'required': required,
        'available': available,
      },
    );
  }

  factory TradingError.invalidSymbol(String symbol) {
    return TradingError(
      code: ErrorCodes.tradingInvalidSymbol,
      message: 'Invalid trading symbol',
      details: {'symbol': symbol},
    );
  }

  factory TradingError.invalidQuantity(double quantity) {
    return TradingError(
      code: ErrorCodes.tradingInvalidQuantity,
      message: 'Invalid quantity for trade',
      details: {'quantity': quantity},
    );
  }
}

/// Authentication-specific error classes
class AuthError extends AppError {
  AuthError({
    required String code,
    required String message,
    Map<String, dynamic>? details,
    String? traceId,
  }) : super(
          code: code,
          message: message,
          details: details,
          traceId: traceId,
        );

  factory AuthError.invalidCredentials() {
    return AuthError(
      code: ErrorCodes.authInvalidCredentials,
      message: 'Invalid credentials provided',
    );
  }

  factory AuthError.tokenExpired() {
    return AuthError(
      code: ErrorCodes.authTokenExpired,
      message: 'Authentication token has expired',
    );
  }

  factory AuthError.invalidSignature() {
    return AuthError(
      code: ErrorCodes.authSignatureInvalid,
      message: 'Invalid signature provided',
    );
  }
}

/// Blockchain-specific error classes
class BlockchainError extends AppError {
  BlockchainError({
    required String code,
    required String message,
    Map<String, dynamic>? details,
    String? traceId,
  }) : super(
          code: code,
          message: message,
          details: details,
          traceId: traceId,
        );

  factory BlockchainError.connectionFailed() {
    return BlockchainError(
      code: ErrorCodes.blockchainConnectionFailed,
      message: 'Failed to connect to blockchain network',
    );
  }

  factory BlockchainError.transactionFailed(String txHash) {
    return BlockchainError(
      code: ErrorCodes.blockchainTransactionFailed,
      message: 'Blockchain transaction failed',
      details: {'transaction_hash': txHash},
    );
  }

  factory BlockchainError.insufficientGas({
    required double required,
    required double available,
  }) {
    return BlockchainError(
      code: ErrorCodes.blockchainInsufficientGas,
      message: 'Insufficient gas for transaction',
      details: {
        'required': required,
        'available': available,
      },
    );
  }
}

File: shared/errors/app_error.py
"""
Application Error Classes - Python implementation
Standardized error handling across Python services
"""

import uuid
from datetime import datetime
from typing import Dict, Any, Optional


def generate_trace_id() -> str:
    """Generate a random trace ID for error tracking"""
    return str(uuid.uuid4()).replace('-', '')[:16]


class AppError(Exception):
    """Application error class with standardized structure"""
    
    def __init__(
        self,
        code: str,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        trace_id: Optional[str] = None
    ):
        self.code = code
        self.message = message
        self.details = details or {}
        self.trace_id = trace_id or generate_trace_id()
        self.timestamp = datetime.now()
        super().__init__(self.message)

    def to_dict(self) -> Dict[str, Any]:
        """Convert error to dictionary format"""
        return {
            'error': {
                'code': self.code,
                'message': self.message,
                'details': self.details,
                'timestamp': self.timestamp.isoformat(),
                'trace_id': self.trace_id,
            }
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'AppError':
        """Create AppError from dictionary"""
        error_data = data['error']
        
        return cls(
            code=error_data['code'],
            message=error_data['message'],
            details=error_data.get('details'),
            trace_id=error_data.get('trace_id'),
        )

    def __str__(self) -> str:
        return f'AppError(code={self.code}, message={self.message}, trace_id={self.trace_id})'

    def __repr__(self) -> str:
        return self.__str__()


class ErrorCodes:
    """Common error codes for the application"""
    
    # Trading errors (001-099)
    TRADING_INSUFFICIENT_BALANCE = 'TRADING_001'
    TRADING_INVALID_SYMBOL = 'TRADING_002'
    TRADING_INVALID_QUANTITY = 'TRADING_003'
    TRADING_INVALID_PRICE = 'TRADING_004'
    TRADING_ORDER_NOT_FOUND = 'TRADING_005'
    TRADING_MARKET_CLOSED = 'TRADING_006'
    TRADING_POSITION_NOT_FOUND = 'TRADING_007'
    TRADING_RISK_LIMIT_EXCEEDED = 'TRADING_008'
    
    # Authentication errors (100-199)
    AUTH_INVALID_CREDENTIALS = 'AUTH_100'
    AUTH_TOKEN_EXPIRED = 'AUTH_101'
    AUTH_TOKEN_INVALID = 'AUTH_102'
    AUTH_PERMISSION_DENIED = 'AUTH_103'
    AUTH_SIGNATURE_INVALID = 'AUTH_104'
    AUTH_NONCE_INVALID = 'AUTH_105'
    
    # Blockchain errors (200-299)
    BLOCKCHAIN_CONNECTION_FAILED = 'BLOCKCHAIN_200'
    BLOCKCHAIN_TRANSACTION_FAILED = 'BLOCKCHAIN_201'
    BLOCKCHAIN_INSUFFICIENT_GAS = 'BLOCKCHAIN_202'
    BLOCKCHAIN_CONTRACT_ERROR = 'BLOCKCHAIN_203'
    BLOCKCHAIN_INVALID_ADDRESS = 'BLOCKCHAIN_204'
    BLOCKCHAIN_NETWORK_ERROR = 'BLOCKCHAIN_205'
    
    # RAG/Knowledge base errors (300-399)
    RAG_INDEX_NOT_FOUND = 'RAG_300'
    RAG_SEARCH_FAILED = 'RAG_301'
    RAG_DOCUMENT_NOT_FOUND = 'RAG_302'
    RAG_PROCESSING_FAILED = 'RAG_303'
    RAG_DATABASE_ERROR = 'RAG_304'
    
    # API/Network errors (400-499)
    API_NETWORK_ERROR = 'API_400'
    API_TIMEOUT_ERROR = 'API_401'
    API_RATE_LIMIT_EXCEEDED = 'API_402'
    API_INVALID_REQUEST = 'API_403'
    API_INVALID_RESPONSE = 'API_404'
    API_SERVICE_UNAVAILABLE = 'API_405'
    
    # System errors (500-599)
    SYSTEM_CONFIGURATION_ERROR = 'SYSTEM_500'
    SYSTEM_DATABASE_ERROR = 'SYSTEM_501'
    SYSTEM_FILE_ERROR = 'SYSTEM_502'
    SYSTEM_MEMORY_ERROR = 'SYSTEM_503'
    SYSTEM_UNKNOWN_ERROR = 'SYSTEM_504'


class TradingError(AppError):
    """Trading-specific error class"""
    
    def __init__(
        self,
        code: str,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        trace_id: Optional[str] = None
    ):
        super().__init__(code, message, details, trace_id)

    @classmethod
    def insufficient_balance(cls, required: float, available: float) -> 'TradingError':
        return cls(
            code=ErrorCodes.TRADING_INSUFFICIENT_BALANCE,
            message='Insufficient balance for trade',
            details={
                'required': required,
                'available': available,
            }
        )

    @classmethod
    def invalid_symbol(cls, symbol: str) -> 'TradingError':
        return cls(
            code=ErrorCodes.TRADING_INVALID_SYMBOL,
            message='Invalid trading symbol',
            details={'symbol': symbol}
        )

    @classmethod
    def invalid_quantity(cls, quantity: float) -> 'TradingError':
        return cls(
            code=ErrorCodes.TRADING_INVALID_QUANTITY,
            message='Invalid quantity for trade',
            details={'quantity': quantity}
        )

    @classmethod
    def order_not_found(cls, order_id: str) -> 'TradingError':
        return cls(
            code=ErrorCodes.TRADING_ORDER_NOT_FOUND,
            message='Order not found',
            details={'order_id': order_id}
        )

    @classmethod
    def market_closed(cls, symbol: str) -> 'TradingError':
        return cls(
            code=ErrorCodes.TRADING_MARKET_CLOSED,
            message='Market is closed for trading',
            details={'symbol': symbol}
        )


class AuthError(AppError):
    """Authentication-specific error class"""
    
    def __init__(
        self,
        code: str,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        trace_id: Optional[str] = None
    ):
        super().__init__(code, message, details, trace_id)

    @classmethod
    def invalid_credentials(cls) -> 'AuthError':
        return cls(
            code=ErrorCodes.AUTH_INVALID_CREDENTIALS,
            message='Invalid credentials provided'
        )

    @classmethod
    def token_expired(cls) -> 'AuthError':
        return cls(
            code=ErrorCodes.AUTH_TOKEN_EXPIRED,
            message='Authentication token has expired'
        )

    @classmethod
    def invalid_signature(cls) -> 'AuthError':
        return cls(
            code=ErrorCodes.AUTH_SIGNATURE_INVALID,
            message='Invalid signature provided'
        )

    @classmethod
    def permission_denied(cls, resource: str) -> 'AuthError':
        return cls(
            code=ErrorCodes.AUTH_PERMISSION_DENIED,
            message='Permission denied for resource',
            details={'resource': resource}
        )


class BlockchainError(AppError):
    """Blockchain-specific error class"""
    
    def __init__(
        self,
        code: str,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        trace_id: Optional[str] = None
    ):
        super().__init__(code, message, details, trace_id)

    @classmethod
    def connection_failed(cls, network: str = 'unknown') -> 'BlockchainError':
        return cls(
            code=ErrorCodes.BLOCKCHAIN_CONNECTION_FAILED,
            message='Failed to connect to blockchain network',
            details={'network': network}
        )

    @classmethod
    def transaction_failed(cls, tx_hash: str, reason: str = 'unknown') -> 'BlockchainError':
        return cls(
            code=ErrorCodes.BLOCKCHAIN_TRANSACTION_FAILED,
            message='Blockchain transaction failed',
            details={
                'transaction_hash': tx_hash,
                'reason': reason
            }
        )

    @classmethod
    def insufficient_gas(cls, required: float, available: float) -> 'BlockchainError':
        return cls(
            code=ErrorCodes.BLOCKCHAIN_INSUFFICIENT_GAS,
            message='Insufficient gas for transaction',
            details={
                'required': required,
                'available': available
            }
        )

    @classmethod
    def contract_error(cls, contract_address: str, error_message: str) -> 'BlockchainError':
        return cls(
            code=ErrorCodes.BLOCKCHAIN_CONTRACT_ERROR,
            message='Smart contract execution failed',
            details={
                'contract_address': contract_address,
                'error_message': error_message
            }
        )


class RagError(AppError):
    """RAG/Knowledge base specific error class"""
    
    def __init__(
        self,
        code: str,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        trace_id: Optional[str] = None
    ):
        super().__init__(code, message, details, trace_id)

    @classmethod
    def search_failed(cls, query: str, reason: str = 'unknown') -> 'RagError':
        return cls(
            code=ErrorCodes.RAG_SEARCH_FAILED,
            message='RAG search failed',
            details={
                'query': query,
                'reason': reason
            }
        )

    @classmethod
    def document_not_found(cls, document_id: str) -> 'RagError':
        return cls(
            code=ErrorCodes.RAG_DOCUMENT_NOT_FOUND,
            message='Document not found in knowledge base',
            details={'document_id': document_id}
        )

    @classmethod
    def processing_failed(cls, operation: str, reason: str = 'unknown') -> 'RagError':
        return cls(
            code=ErrorCodes.RAG_PROCESSING_FAILED,
            message=f'RAG processing failed for operation: {operation}',
            details={
                'operation': operation,
                'reason': reason
            }
        )

    @classmethod
    def database_error(cls, operation: str, error_message: str) -> 'RagError':
        return cls(
            code=ErrorCodes.RAG_DATABASE_ERROR,
            message='RAG database operation failed',
            details={
                'operation': operation,
                'error_message': error_message
            }
        )